db/commitlog/commitlog.cc

```diff
@@ -1853,10 +1853,13 @@ future<> db::commitlog::segment_manager::oversized_allocation(entry_writer& writ
             }
         }
     }
-    assert(_request_controller.available_units() == 0);
-
+    SCYLLA_ASSERT(_request_controller.available_units() == 0);
+    SCYLLA_ASSERT(permit.count() == max_request_controller_units());
+    auto nw = _request_controller.waiters();
     permit.return_all();
-    assert(_request_controller.available_units() == ssize_t(max_request_controller_units()));
+    // #20633 cannot guarantee controller avail is now full, since we could have had waiters when doing
+    // return all -> now will be less avail
+    SCYLLA_ASSERT(nw > 0 || _request_controller.available_units() == ssize_t(max_request_controller_units()));
```

Comment:
```suggestion
    SCYLLA_ASSERT(nw + _request_controller.available_units() == ssize_t(max_request_controller_units()));
```
Isn't this better represent the state we expect?

Reply:
nw == number of waiting tasks. Not amount of waited units.

Reply:
And each task may use a few units?

Reply:
Each task uses as many units as base allocation bytes.